### ISSUE TITLE
conf option questions

### ISSUE THREAD

**Issue by Blisto91** [doitsujin/dxvk#2582]
Hi there. I hope it's alright i open a quick issue to ask a few questions (can close it again quickly), was not sure where else i could ask.

When testing games and going through the issues i sometimes mess with the conf options and was wondering about a few things, mainly about the ones relating to vram/memory.
I've read a bit of your comments from old issues and peaked slightly at some code, but lack some of the knowledge to dig deeper and understand.

As an example for a dx9 game what is the practical difference between setting `dxgi.maxDeviceMemory` `dxgi.maxSharedMemory` & `d3d9.maxAvailableMemory`. Mostly the first and last since i guess `maxSharedMemory` is just how much regular memory is reported to be allowed to be used as vram? Tho how does this relate to a APU/igpu such as the steam deck?
`maxDeviceMemory` i would guess would be like saying what actual memory i want to report my gpu having? 
`maxAvailableMemory` i gather has something to do with initial memory amount allowed to be dedicated? Although i'm not sure here.

Would it even make sense to set `maxAvailableMemory` to more than 4096? I see it's used in a place where it's multiplied with `Megabytes` which is 1024 * 1024. So if the variable is a 32bit uint it can't get much larger anyway?

Lastly i saw a couple of issues where `d3d9.memoryTrackTest` solved the problem. What is this used for? Tracking and freeing vram?

Thx for any replies.

**Comment by Deus-nsf** [comment]
I personally tried to play around d3d9.maxAvailableMemory (to circumvent the out of memory issue with black mesa at xen) but any attempts at changing the value did either nothing, or increased crashes. Which makes sense I suppose but I tried anyways.

I'd be curious with the d3d9.memoryTrackTest though, didn't try that one yet.

**Comment by Blisto91** [comment]
Depends on how f it's VRAM or system memory issues i guess. If it's the latter this PR would be a better bet.
https://github.com/doitsujin/dxvk/pull/2524

**Comment by Deus-nsf** [comment]
yep BM is 32 bit, and playing with d3d9.memoryTrackTest didn't help as i suspected

**Comment by K0bin** [comment]
> As an example for a dx9 game what is the practical difference between setting dxgi.maxDeviceMemory dxgi.maxSharedMemory & d3d9.maxAvailableMemory.

D3D9 games don't use dxgi, so only the d3d9 option will have an effect on those. As you said, it only changes what DXVK tells the game when it asks for how much memory is available.

DXVK also counts the amount of memory used for D3DPOOL_MANAGED resources and refused to create new ones once that has the maximum amount of memory we report. Some games rely on that and keep creating resources until it fails. That is however not something you'd want for heavily modded games, hence why we have an option to turn that behavior off.

**Comment by Blisto91** [comment]
Thanks for the reply K0bin. What about the memoryTrackTest?

@Deus-nsf I can try to test the game and see if i can reproduce

**Comment by Deus-nsf** [comment]
no it's fine, seems like a lost cause for now, the command helped but it still crashes, I am also guilty of using a few mods from the workshop, which probably doesn't help with the memory load (even though it doesn't crash in D3D9)

**Comment by Blisto91** [comment]
You can't stop me hohoho. 
I've been meaning to play through black mesa again after they added Zen.

**Comment by Deus-nsf** [comment]
okay, a very interesting fact, on windows (useful to directly compared D3D9 and Vulkan, when it works of course) on D3D9 my RAM usage goes from 6.5 GB to 8.5 GB, with Vulkan through the dxvk dll, it's 9.5 GB.

if you want to test something, please test that as it may prove very interesting. if you don't have windows just load xen and compare your RAM usage before launching the game, and after loading a new game from xen (sv_unlockedchapters 19 from the console if it's locked)

**Comment by Blisto91** [comment]
Roger. I have windows so can test that.
We should probably move this to your other or a new dxvk issue tho. Or proton issue hehe

**Comment by Deus-nsf** [comment]
a new issue then, but I need to test previous dxvk versions as well first, I'm not sure they use to allocate so much ram, but I dont know.

**Comment by K0bin** [comment]
@Blisto91 I should've been a bit more clear about this. `d3d9.memoryTrackTest` controls whether we fail resource creation based on tracked memory usage. I've edited the comment to clarify that.

@Deus-nsf Yes, please open a new issue for Black Mesa please. I'm playing through that again anyway right now so I'll probably look into that soon-ish.

I answered the question so I'm gonna close this. Feel free to ask if there's still things that aren't clear.

**Comment by Blisto91** [comment]
Thanks K0bin. I think i get it.
Am i correct in understanding that when you set `d3d9.memoryTrackTest` to true is when dxvk will refuse to create more once that max reported has been reached? Like i see it's been set to true for some games to "Prevents the game from consuming all system memory." in the case of Anarchy Online

**Comment by K0bin** [comment]
Exactly.

**Comment by Blisto91** [comment]
Quick question that isn't about the conf but is about memory.
What is it that "vidmem heap 2" shows in the hud? I'm guessing vidmem heap 0 is regular old vram.

**Comment by Blisto91** [comment]
@Deus-nsf just as a quick fyi i tried a play-through of the Xen levels of the windows version on linux through proton & dxvk. But i didn't experience a crash.

**Comment by Deus-nsf** [comment]
@Blisto91 Sorry about that, one of the biggest reasons of the crashing was identified after a long and painful process of troubleshooting, it's not "just" that, but a big reason was the "Enable Shader Pre-Caching" and ESPECIALLY the "Allow background processing of Vulkan shaders" option on the Steam Shader Pre-Caching options.
Really no idea why, but as soon as I unchecked those options, memory utilization was down 1 whole GB! And the crashing was significantly less occurent as a result.
